### PR TITLE
Support parsing performance data

### PR DESCRIPTION
**Pull request by stephen-soltesz**
This PR adds support for parsing and exporting Nagios performance data as well as a mode for dumping all known metrics to stdout and then exiting.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/prometheus-nagios-exporter/2)

### CODE REVIEW

File: nagios_exporter.py [hunk 1]
```diff
@@ -49,20 +91,43 @@ class NagiosConnectError(NagiosError):
...
     parser.add_argument(
-        '--path', help='Absolute path to livestatus Nagios UNIX socket.')
+        '--path', default='/var/lib/nagios3/rw/livestatus',
+        help='Absolute path to livestatus Nagios UNIX socket.')
 
     # TODO: choose appropriate port based on:
     # https://github.com/prometheus/prometheus/wiki/Default-port-allocations
     parser.add_argument(
-        '--port', type=int, default=5000,
+        '--port', type=int, default=5000, metavar='5000',
```
Comment (pboothe) [inline nagios_exporter.py:102]: Should be `metavar='PORT'` probably.

File: nagios_exporter.py [hunk 2]
```diff
@@ -49,20 +91,43 @@ class NagiosConnectError(NagiosError):
...
         help='Export server should listen on port.')
 
-    # TODO: support --whitelist patterns to limit exported services.
+    # Report metrics in the "--whitelist" or "--all_metrics".
+    parser.add_argument(
+        '--whitelist', default=None, action='append', metavar='<pattern>',
+        help=('Default: only export metrics for services that include this '
+              'whitelist pattern. Can be specified multiple times.'))
```
Comment (pboothe) [inline nagios_exporter.py]: Will the patterns be combined with OR or AND?

File: nagios_exporter.py [hunk 3]
```diff
@@ -148,6 +213,137 @@ def _receive(self, count):
...
+    return value, unit
+
+
+def convert_value_to_base_unit(value, unit):
+    """Converts value to canonical units."""
+    if not unit:
+        return value
+
+    if unit not in UNIT_TO_SCALE:
+        # TODO: log missing unit.
```
Comment (pboothe) [inline nagios_exporter.py]: Replace TODO with `logging.warning('Unknown unit: %s', unit)`

File: nagios_exporter.py [hunk 4]
```diff
@@ -235,16 +433,23 @@ def collect_metrics(args, lines):
...
-        lines.extend(get_services(session))
+    if args.whitelist or args.all_metrics or args.dump_metrics:
+        with contextlib.closing(connect(args.path)) as sock:
+            session = LiveStatus(sock)
+            services = get_services(
+                session, args.use_perf_data, args.data_names)
+
+        if args.whitelist:
+            for whitelist in args.whitelist:
+                lines.extend(filter(lambda x: whitelist in x, services))
```
Comment (pboothe) [inline nagios_exporter.py]: This could potentially put items from `service` in `lines` multiple times. Is this okay?

**Review by pboothe — APPROVED** [review]
Small comments, but nothing blocking. LGTM.

**Comment by stephen-soltesz** [comment]
Thank you for finding the bug! PTAL to see if there's a simpler way to uniquely select matching patterns that preserves the original metric order?

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[nagios_exporter.py, line 102 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2#-KmWxLDusKpvUwBUK0MB-r2-102:-KmX9mRybkxQANEJg3k9:b3kitgr) ([raw file](https://github.com/m-lab/prometheus-nagios-exporter/blob/3807df786bd3bb1b2077621e0a0153e2806e6445/nagios_exporter.py#L102)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Should be `metavar='PORT'` probably.
</blockquote></details>

That's the default, but I felt it would be more helpful to show the default value. But, I don't do that consistently.  ¯\_(ツ)_/¯

---

*[nagios_exporter.py, line 109 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2#-KmWxLDusKpvUwBUK0MB-r2-109:-KmXAkkvZCbGzjfkdH9S:b-4qq70b) ([raw file](https://github.com/m-lab/prometheus-nagios-exporter/blob/3807df786bd3bb1b2077621e0a0153e2806e6445/nagios_exporter.py#L109)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Will the patterns be combined with OR or AND?
</blockquote></details>

OR is the intention, but it looks like the whitelist pattern could originally match the same metrics multiple times. That was a bug.

---

*[nagios_exporter.py, line 335 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2#-KmWxLDusKpvUwBUK0MB-r2-335:-KmXBDoP2WtWIpdsXwfW:b-896fix) ([raw file](https://github.com/m-lab/prometheus-nagios-exporter/blob/3807df786bd3bb1b2077621e0a0153e2806e6445/nagios_exporter.py#L335)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Replace TODO with `logging.warning('Unknown unit: %s', unit)`
</blockquote></details>

Done.

---

*[nagios_exporter.py, line 444 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2#-KmWxLDusKpvUwBUK0MB-r2-444:-KmXC1ey71LEGSwawI9U:bfr3ihh) ([raw file](https://github.com/m-lab/prometheus-nagios-exporter/blob/3807df786bd3bb1b2077621e0a0153e2806e6445/nagios_exporter.py#L444)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

This could potentially put items from `service` in `lines` multiple times. Is this okay?
</blockquote></details>

Thank you for catching that. That is not okay. That is a bug. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2#-:-KmXHdX8DRsKvZSCZJoi:b5c75q2)*

File: nagios_exporter.py [hunk 5]
```diff
@@ -235,16 +434,28 @@ def collect_metrics(args, lines):
...
-        lines.extend(get_services(session))
+    if args.whitelist or args.all_metrics or args.dump_metrics:
+        with contextlib.closing(connect(args.path)) as sock:
+            session = LiveStatus(sock)
+            services = get_services(
+                session, args.use_perf_data, args.data_names)
+
+        if args.whitelist:
+            for metric in services:
+                for whitelist in args.whitelist:
```
Comment (pboothe) [inline nagios_exporter.py]: could also be:

```python
if any((whitelist in metric) for whitelist in args.whitelist):
   lines.append(metric)
```

**Review by pboothe — APPROVED** [review]
Still LGTM with an optional alternate implementation suggestion.

**Comment by stephen-soltesz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[nagios_exporter.py, line 445 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2#-KmWxLDusKpvUwBUK0MB-r3-445:-KmXLvvQdXlQktXDKmnU:b-nfiiry) ([raw file](https://github.com/m-lab/prometheus-nagios-exporter/blob/c2d9068bcfd9e41c64b6c15dfc60b73878c9b7f8/nagios_exporter.py#L445)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

could also be:

```python
if any((whitelist in metric) for whitelist in args.whitelist):
   lines.append(metric)
```
</blockquote></details>

That's the one!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-nagios-exporter/2)*

**Pull request merged**